File: Components/DevExpressVCL/Demos/ExpressQuantumGrid/Delphi/StylesMultiDemo/StylesMultiDemoMain.pas
unit StylesMultiDemoMain;

interface

uses
  Windows, Messages, Forms, SysUtils, cxStyles, cxCustomData, cxGraphics, cxFilter,
  cxEdit, DB, cxDBData, Dialogs, cxGridCustomPopupMenu, cxGridPopupMenu,
  Classes, ActnList, ImgList, Controls, Menus, StdCtrls, ExtCtrls,
  cxButtons, cxGridLevel, cxGridCustomTableView, cxGridTableView, cxData,
  cxGridDBTableView, cxClasses, cxControls, cxGridCustomView, cxGrid,
  ComCtrls, cxContainer, cxRadioGroup, cxListBox, cxDataStorage,
  cxLookAndFeelPainters, cxGroupBox, cxLookAndFeels, cxGridStyleSheetsPreview,
  BaseForm, cxGridCardView, cxTextEdit, cxMaskEdit, cxDropDownEdit;

type
  TStylesMultiDemoMainForm = class(TfmBaseForm)
    tvProjects: TcxGridDBTableView;
    lvProjects: TcxGridLevel;
    cxGrid: TcxGrid;
    pnlLeft: TPanel;
    tvProjectsID: TcxGridDBColumn;
    tvProjectsNAME: TcxGridDBColumn;
    tvProjectsMANAGERID: TcxGridDBColumn;
    lvTeam: TcxGridLevel;
    tvTeam: TcxGridDBTableView;
    tvTeamPROJECTID: TcxGridDBColumn;
    tvTeamUSERID: TcxGridDBColumn;
    tvTeamFUNCTION: TcxGridDBColumn;
    Splitter: TSplitter;
    cxGridPopupMenu1: TcxGridPopupMenu;
    OpenDialog: TOpenDialog;
    SaveDialog: TSaveDialog;
    gbUserDefined: TcxGroupBox;
    RadioGroup: TcxRadioGroup;
    gbPredefined: TcxGroupBox;
    cbUserStyleSheets: TcxComboBox;
    btnLoad: TcxButton;
    btnSave: TcxButton;
    lbPredefinedStyleSheets: TcxListBox;
    btnEdit: TcxButton;
    pnlCurrentStyleSheet: TPanel;
    procedure FormCreate(Sender: TObject);
    procedure btnSaveClick(Sender: TObject);
    procedure btnLoadClick(Sender: TObject);
    procedure RadioGroupClick(Sender: TObject);
    procedure cbUserStyleSheetsChange(Sender: TObject);
    procedure lbPredefinedStyleSheetsClick(Sender: TObject);
    procedure btnEditClick(Sender: TObject);
    procedure FormActivate(Sender: TObject);
  private
    function GetCurrentStyleSheet: TcxGridTableViewStyleSheet;
    procedure CreateUserStyleSheetsList;
    procedure CreatePredefinedStyleSheetsList;
    procedure UpdateGridStyleSheets(const AStyleSheet: TcxGridTableViewStyleSheet);
    procedure ChangeVisibility(AType: Integer);
    procedure SetPredefinedStyleSheets;
    procedure SetUserDefinedStyleSheets;
    procedure ClearUserDefinedStyleSheets;
    procedure LoadUserDefinedStyleSheets(AFileName: TFileName);
    procedure SaveUserDefinedStyleSheets(AFileName: TFileName);
  end;

var
  StylesMultiDemoMainForm: TStylesMultiDemoMainForm;

implementation

{$R *.dfm}

uses
  StylesMultiDemoData, cxStyleSheetEditor, AboutDemoForm;

const
  cNone = 0;
  cPredefined = 1;
  cUserDefined = 2;

procedure TStylesMultiDemoMainForm.FormCreate(Sender: TObject);
begin
  CreateUserStyleSheetsList;
  CreatePredefinedStyleSheetsList;
  SetPredefinedStyleSheets;
end;

procedure TStylesMultiDemoMainForm.CreateUserStyleSheetsList;
var
  I: Integer;
begin
  with StylesMultiDemoMainDM.strepUserDefined do
  begin
    cbUserStyleSheets.Properties.Items.Clear;
    for I := 0 to StyleSheetCount - 1 do
      cbUserStyleSheets.Properties.Items.AddObject(StyleSheets[I].Caption, StyleSheets[I]);
    cbUserStyleSheets.ItemIndex := 0;
  end;
end;

procedure TStylesMultiDemoMainForm.CreatePredefinedStyleSheetsList;
var
  I: Integer;
begin
  with StylesMultiDemoMainDM.strepPredefined do
  begin
    lbPredefinedStyleSheets.Clear;
    for I := 0 to StyleSheetCount - 1 do
      lbPredefinedStyleSheets.Items.AddObject(StyleSheets[I].Caption, StyleSheets[I]);
    lbPredefinedStyleSheets.ItemIndex := 0;
  end;
end;

procedure TStylesMultiDemoMainForm.UpdateGridStyleSheets(const AStyleSheet: TcxGridTableViewStyleSheet);

  procedure UpdateView(const AView: TcxGridDBTableView);
  begin
    with AView do
    begin
      BeginUpdate;
      Styles.StyleSheet := AStyleSheet;
      EndUpdate;
    end;
  end;

begin
  if GetCurrentStyleSheet = AStyleSheet then
    Exit;
  UpdateView(tvProjects);
  UpdateView(tvTeam);
  tvProjects.DataController.ClearDetails;  // refresh detail level

  if AStyleSheet <> nil then
    pnlCurrentStyleSheet.Caption := AStyleSheet.Caption
  else
    pnlCurrentStyleSheet.Caption := '';
end;

procedure TStylesMultiDemoMainForm.btnSaveClick(Sender: TObject);
begin
  with SaveDialog do
    if Execute then
      SaveUserDefinedStyleSheets(FileName);
end;

procedure TStylesMultiDemoMainForm.btnLoadClick(Sender: TObject);
begin
  with OpenDialog do
    if Execute then
      LoadUserDefinedStyleSheets(FileName);
end;

procedure TStylesMultiDemoMainForm.RadioGroupClick(Sender: TObject);
begin
  case TcxRadioGroup(Sender).ItemIndex of
    cNone:
      UpdateGridStyleSheets(nil);
    cPredefined:
      SetPredefinedStyleSheets;
    cUserDefined:
      SetUserDefinedStyleSheets;
  end;
  ChangeVisibility(TcxRadioGroup(Sender).ItemIndex);
end;

procedure TStylesMultiDemoMainForm.SetUserDefinedStyleSheets;
begin
  with cbUserStyleSheets.Properties do
    if Items.Count > 0 then
      UpdateGridStyleSheets(TcxGridTableViewStyleSheet(
        Items.Objects[cbUserStyleSheets.ItemIndex]));
end;

procedure TStylesMultiDemoMainForm.SetPredefinedStyleSheets;
begin
  with lbPredefinedStyleSheets do
    if Items.Count > 0 then
      UpdateGridStyleSheets(TcxGridTableViewStyleSheet(Items.Objects[ItemIndex]));
end;

procedure TStylesMultiDemoMainForm.ChangeVisibility(AType: Integer);
begin
  cbUserStyleSheets.Enabled := AType = cUserDefined;
  gbUserDefined.Enabled := AType = cUserDefined;
  btnEdit.Enabled := AType = cUserDefined;
  btnLoad.Enabled := AType = cUserDefined;
  btnSave.Enabled := AType = cUserDefined;

  lbPredefinedStyleSheets.Enabled := AType = cPredefined;
end;

procedure TStylesMultiDemoMainForm.cbUserStyleSheetsChange(
  Sender: TObject);
begin
  SetUserDefinedStyleSheets;
end;

procedure TStylesMultiDemoMainForm.lbPredefinedStyleSheetsClick(
  Sender: TObject);
begin
  SetPredefinedStyleSheets;
end;

procedure TStylesMultiDemoMainForm.btnEditClick(Sender: TObject);
begin
  with cbUserStyleSheets.Properties do
    ShowcxStyleSheetEditor(TcxGridTableViewStyleSheet(
      Items.Objects[cbUserStyleSheets.ItemIndex]), nil);
end;

function TStylesMultiDemoMainForm.GetCurrentStyleSheet: TcxGridTableViewStyleSheet;
begin
  Result := TcxGridTableViewStyleSheet(tvProjects.Styles.StyleSheet);
end;

procedure TStylesMultiDemoMainForm.LoadUserDefinedStyleSheets(AFileName: TFileName);
begin
  UpdateGridStyleSheets(nil);
  ClearUserDefinedStyleSheets;

  LoadStyleSheetsFromIniFile(AFileName, StylesMultiDemoMainDM.strepUserDefined,
    TcxGridTableViewStyleSheet);

  CreateUserStyleSheetsList;
  SetUserDefinedStyleSheets;
end;

procedure TStylesMultiDemoMainForm.SaveUserDefinedStyleSheets(AFileName: TFileName);
var
  AList: TList;
begin
  AList := TList.Create;
  try
    PopulateStyleSheetsList(AList);
    SaveStyleSheetsToIniFile(AFileName, AList);
  finally
    AList.Free;
  end;
end;

procedure TStylesMultiDemoMainForm.ClearUserDefinedStyleSheets;
begin
  with StylesMultiDemoMainDM.strepUserDefined do
  begin
    Clear;
    ClearStyleSheets;
  end;
end;

procedure TStylesMultiDemoMainForm.FormActivate(Sender: TObject);
begin
  OpenDialog.InitialDir := ExtractFileDir(Application.ExeName);
  SaveDialog.InitialDir := OpenDialog.InitialDir;
end;

end.

File: Components/DevExpressVCL/Demos/ExpressQuantumGrid/CBuilder/StylesMultiDemo/StylesMultiDemoMain.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#include "shellapi.hpp"
#pragma hdrstop

#include "StylesMultiDemoMain.h"
#include "StylesMultiDemoData.h"
#include "AboutDemoForm.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma link "cxClasses"
#pragma link "cxControls"
#pragma link "cxCustomData"
#pragma link "cxData"
#pragma link "cxDBData"
#pragma link "cxEdit"
#pragma link "cxFilter"
#pragma link "cxGraphics"
#pragma link "cxGrid"
#pragma link "cxGridCustomTableView"
#pragma link "cxGridCustomView"
#pragma link "cxGridDBTableView"
#pragma link "cxGridLevel"
#pragma link "cxGridTableView"
#pragma link "cxStyles"
#pragma link "cxButtons"
#pragma link "cxGridCustomPopupMenu"
#pragma link "cxGridPopupMenu"
#pragma link "cxLookAndFeelPainters"
#pragma link "cxStyleSheetEditor"
#pragma link "cxRadioGroup"
#pragma link "cxListBox"
#pragma link "cxContainer"
#pragma link "cxGroupBox"
#pragma link "cxGridStyleSheetsPreview"
#pragma link "cxDataStorage"
#pragma link "cxDropDownEdit"
#pragma link "cxGridCardView"
#pragma link "cxMaskEdit"
#pragma link "cxTextEdit"
#pragma resource "*.dfm"
TStylesMultiDemoMainForm *StylesMultiDemoMainForm;

const int cNone = 0;
const int cPredefined = 1;
const int cUserDefined = 2;

//---------------------------------------------------------------------------
__fastcall TStylesMultiDemoMainForm::TStylesMultiDemoMainForm(TComponent* Owner)
  : TfmBaseForm(Owner)
{
}
//---------------------------------------------------------------------------

void __fastcall TStylesMultiDemoMainForm::FormCreate(TObject *Sender)
{
  CreateUserStyleSheetsList();
  CreatePredefinedStyleSheetsList();
  SetPredefinedStyleSheets();
}
//---------------------------------------------------------------------------

void __fastcall TStylesMultiDemoMainForm::btnSaveClick(TObject *Sender)
{
    if (SaveDialog->Execute())
      SaveUserDefinedStyleSheets(SaveDialog->FileName);
}
//---------------------------------------------------------------------------

void __fastcall TStylesMultiDemoMainForm::btnLoadClick(TObject *Sender)
{
  if (OpenDialog->Execute())
    LoadUserDefinedStyleSheets(OpenDialog->FileName);
}
//---------------------------------------------------------------------------

void __fastcall TStylesMultiDemoMainForm::RadioGroupClick(TObject *Sender)
{
  switch (((TcxRadioGroup*)Sender)->ItemIndex) {
    case cNone:
      UpdateGridStyleSheets(NULL);
    case cPredefined:
      SetPredefinedStyleSheets();
    case cUserDefined:
      SetUserDefinedStyleSheets();
  }
  ChangeVisibility(((TcxRadioGroup*)Sender)->ItemIndex);
}
//---------------------------------------------------------------------------

void __fastcall TStylesMultiDemoMainForm::cbUserStyleSheetsChange(TObject *Sender)
{
  SetUserDefinedStyleSheets();
}
//---------------------------------------------------------------------------

void __fastcall TStylesMultiDemoMainForm::lbPredefinedStyleSheetsClick(TObject *Sender)
{
  SetPredefinedStyleSheets();
}
//---------------------------------------------------------------------------

void __fastcall TStylesMultiDemoMainForm::btnEditClick(TObject *Sender)
{
  ShowcxStyleSheetEditor((TcxGridTableViewStyleSheet*)(cbUserStyleSheets->
  	Properties->Items->Objects[cbUserStyleSheets->ItemIndex]), NULL);
}
//---------------------------------------------------------------------------

void __fastcall TStylesMultiDemoMainForm::FormActivate(TObject *Sender)
{
  OpenDialog->InitialDir = ExtractFileDir(Application->ExeName);
  SaveDialog->InitialDir = OpenDialog->InitialDir;
}
//---------------------------------------------------------------------------

TcxGridTableViewStyleSheet* TStylesMultiDemoMainForm::GetCurrentStyleSheet()
{
  return (TcxGridTableViewStyleSheet*)tvProjects->Styles->StyleSheet;
}
//---------------------------------------------------------------------------

void TStylesMultiDemoMainForm::CreateUserStyleSheetsList()
{
  cbUserStyleSheets->Properties->Items->Clear();
  for (int I = 0 ; I < StylesMultiDemoMainDM->strepUserDefined->StyleSheetCount; I++)
    cbUserStyleSheets->Properties->Items->AddObject(
	  StylesMultiDemoMainDM->strepUserDefined->StyleSheets[I]->Caption,
      StylesMultiDemoMainDM->strepUserDefined->StyleSheets[I]);
  cbUserStyleSheets->ItemIndex = 0;
}
//---------------------------------------------------------------------------

void TStylesMultiDemoMainForm::CreatePredefinedStyleSheetsList()
{
  lbPredefinedStyleSheets->Clear();
  for (int I = 0; I < StylesMultiDemoMainDM->strepPredefined->StyleSheetCount; I++)
	lbPredefinedStyleSheets->Items->AddObject(
	  StylesMultiDemoMainDM->strepPredefined->StyleSheets[I]->Caption,
      StylesMultiDemoMainDM->strepPredefined->StyleSheets[I]);
  lbPredefinedStyleSheets->ItemIndex = 0;
}
//---------------------------------------------------------------------------

void TStylesMultiDemoMainForm::UpdateView(TcxGridDBTableView *AView,
  TcxGridTableViewStyleSheet *AStyleSheet)
{
  AView->BeginUpdate();
  AView->Styles->StyleSheet = AStyleSheet;
  AView->EndUpdate();
}
//---------------------------------------------------------------------------

void TStylesMultiDemoMainForm::UpdateGridStyleSheets(TcxGridTableViewStyleSheet *AStyleSheet)
{
  if (GetCurrentStyleSheet() == AStyleSheet) return;
  UpdateView(tvProjects, AStyleSheet);
  UpdateView(tvTeam, AStyleSheet);
  tvProjects->DataController->ClearDetails();  // refresh detail level

  if (AStyleSheet != NULL)
    pnlCurrentStyleSheet->Caption = AStyleSheet->Caption;
  else
    pnlCurrentStyleSheet->Caption = "";
}
//---------------------------------------------------------------------------

void TStylesMultiDemoMainForm::ChangeVisibility(int AType)
{
  cbUserStyleSheets->Enabled = (AType == cUserDefined);
  gbUserDefined->Enabled = (AType == cUserDefined);
  btnEdit->Enabled = (AType == cUserDefined);
  btnLoad->Enabled = (AType == cUserDefined);
  btnSave->Enabled = (AType == cUserDefined);

  lbPredefinedStyleSheets->Enabled = (AType == cPredefined);
}
//---------------------------------------------------------------------------

void TStylesMultiDemoMainForm::SetPredefinedStyleSheets()
{
  if (lbPredefinedStyleSheets->Items->Count > 0)
    UpdateGridStyleSheets(
    (TcxGridTableViewStyleSheet*)lbPredefinedStyleSheets->Items->
      Objects[lbPredefinedStyleSheets->ItemIndex]);
}
//---------------------------------------------------------------------------

void TStylesMultiDemoMainForm::SetUserDefinedStyleSheets()
{
  if (cbUserStyleSheets->Properties->Items->Count > 0)
	UpdateGridStyleSheets((TcxGridTableViewStyleSheet*)cbUserStyleSheets->
	  Properties->Items->Objects[cbUserStyleSheets->ItemIndex]);
}
//---------------------------------------------------------------------------

void TStylesMultiDemoMainForm::ClearUserDefinedStyleSheets()
{
  StylesMultiDemoMainDM->strepUserDefined->Clear();
  StylesMultiDemoMainDM->strepUserDefined->ClearStyleSheets();
}
//---------------------------------------------------------------------------

void TStylesMultiDemoMainForm::LoadUserDefinedStyleSheets(TFileName AFileName)
{
  UpdateGridStyleSheets(NULL);
  ClearUserDefinedStyleSheets();

  LoadStyleSheetsFromIniFile(AFileName, StylesMultiDemoMainDM->strepUserDefined,
    __classid(TcxGridTableViewStyleSheet), NULL, NULL, NULL, NULL);

  CreateUserStyleSheetsList();
  SetUserDefinedStyleSheets();
}
//---------------------------------------------------------------------------

void TStylesMultiDemoMainForm::SaveUserDefinedStyleSheets(TFileName AFileName)
{
  TList *AList = new TList();
  try {
    StylesMultiDemoMainDM->PopulateStyleSheetsList(AList);
    SaveStyleSheetsToIniFile(AFileName, AList);
  }
  __finally {
    delete AList;
  }
}
//---------------------------------------------------------------------------

